src/Common/CurrentMemoryTracker.cpp

```diff
@@ -51,15 +51,23 @@ void CurrentMemoryTracker::allocImpl(Int64 size, bool throw_if_memory_exceeded)
     {
         if (current_thread)
         {
-            current_thread->untracked_memory += size;
+            Int64 will_be = current_thread->untracked_memory + size;
+            Int64 limit = current_thread->untracked_memory_limit + current_thread->untracked_memory_limit_increase;
 
-            if (current_thread->untracked_memory > current_thread->untracked_memory_limit)
+            if (will_be > limit)
             {
-                /// Zero untracked before track. If tracker throws out-of-limit we would be able to alloc up to untracked_memory_limit bytes
+                /// Increase limit before track. If tracker throws out-of-limit we would be able to alloc up to untracked_memory_limit bytes
                 /// more. It could be useful to enlarge Exception message in rethrow logic.
-                Int64 tmp = current_thread->untracked_memory;
+                current_thread->untracked_memory_limit_increase = current_thread->untracked_memory_limit;
```

Comment:
@KochetovNicolai found and interesting case here - https://github.com/ClickHouse/ClickHouse/pull/40424#issuecomment-1221416941

Reply:
But I'm suggesting #40321 anyway